src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js

```diff
@@ -1,171 +1,594 @@
-                .on("end", updateChart)
+            scaleAndAxis.xScale.domain([xMin,xMax]);
+            scaleAndAxis.yScale.domain([yMin,yMax]);
 
-            var scatter = svg.append('g')
-                .attr("clip-path", "url(#clip)")
+            self.zoomingHelper(scaleAndAxis,box_width,scatter,points);
 
-            scatter
-                .selectAll("circle")
-                .data(_data)
-                .enter()
-                .append("circle")
-                .attr("cx", function (d) {
-                    return x(d[headerNames[0]]);
-                })
-                .attr("cy", function (d) {
-                    return y(d[headerNames[1]]);
-                })
-                .attr("r", 8)
-                .style("fill", function (d) {
-                    return color(d[headerNames[2]])
-                })
-                .style("opacity", 0.5)
-
-            scatter
-                .append("g")
-                .attr("class", "brush")
-                .call(brush);
-
-            var idleTimeout
-
-            function idled() {
-                idleTimeout = null;
+            brushElem.call(brush.move,null);
+        }
+
+        const zoomInKeyEvent = (event) => {
+            if (shortcuts.zoomIn(event)) {
+                zoomIn();
+                selectedZoomBtn.style.display = "none";
             }
+        };
 
-            function updateChart() {
-                let extent = d3.event.selection;
-
-                if (!extent) {
-                    if (!idleTimeout) return idleTimeout = setTimeout(idled, 350);
-                    x.domain([4, 8])
-                } else {
-                    x.domain([x.invert(extent[0]), x.invert(extent[1])])
-                    scatter.select(".brush").call(brush.move, null)
-                }
-
-                xAxis.transition().duration(1000).call(d3.axisBottom(x))
-                scatter
-                    .selectAll("circle")
-                    .transition().duration(1000)
-                    .attr("cx", function (d) {
-                        return x(d[headerNames[0]]);
-                    })
-                    .attr("cy", function (d) {
-                        return y(d[headerNames[1]]);
-                    })
+        /**
+         * Updates plot when brushing.
+         */
+        function updateChart() {
+            let selectionEvent = d3.event.selection;
+            selectedZoomBtn.style.display = "inline-block";
+            selectedZoomBtn.addEventListener("click",zoomIn,true)
+            document.addEventListener('keydown',zoomInKeyEvent,true);
+            extent = selectionEvent;
+        }
 
-            }
+        /**
+         * Removes brush, keyboard event and zoom button when end event is captured.
+         */
+        const endBrushing = function (_) {
+            brushElem.call(brush.move,null);
+            selectedZoomBtn.style.display = "none";
+            selectedZoomBtn.removeEventListener("click",zoomIn,true)
```

Comment:
these lines are duplicated in this codebase, while they just mean "hideZoomButton". They should be refactored to separate function. The same applies to hiding / showing other buttons if there is other duplicated code